src/libcore/mem/maybe_uninit.rs

```diff
@@ -402,6 +402,13 @@ impl<T> MaybeUninit<T> {
     ///
     /// [inv]: #initialization-invariant
     ///
+    /// On top of that, remember that most types have additional invariants beyond merely
+    /// being considered initialized at the type level. For example, a `1`-initialized [`Vec<T>`]
+    /// is considered initialized because the only requirement the compiler knows about it
+    /// is that the data pointer must be non-null. Creating such a `Vec<T>` does not cause
```

Comment:
Is this documented anywhere else? I'm not sure we want to settle on `Vec` being non-null as the only requirement.

Reply:
This is a copy of text we already have in the `MaybeUninit` type-level docs.

I agree with your sentiment, not sure what would make a better example though.

Reply:
Yeah, I can't think of anything simple either. Maybe we could say creating `&u8` from `1` instead? That seems somehow simpler and doesn't get into `Vec` which is a pretty complex type from the user's point of view (generics, large API, etc.) even if all that doesn't matter for this case.

Reply:
Maybe use an example that is more domain specific? e.g. something involving a "proof of work" token that if misused triggers UB? //handwave-handwave

Reply:
I think that would have to involve showing some code etc.

I was more thinking if we should weasle a bit more? Like, "with the current implementation, the all-1 bit pattern is legal (but this does not constitute a stable guarantee)".

Reply:
Yeah I suppose that could work.

If you want something short-ish there's:

```rust
pub struct Id<S, T>(core::marker::PhantomData<(fn(S) -> S, fn(T) -> T)>);

pub fn refl<T>() -> Id<T, T> { Id(core::marker::PhantomData) }

impl<S, T> Id<S, T> {
    /// Casts a value of type `S` to `T`.
    pub fn cast(self, value: S) -> T {
        unsafe {
            // This is safe since we know by construction that
            // S == T (including lifetime invariance) always holds.
            let cast_value = core::mem::transmute_copy(&value);
            core::mem::forget(value);
            cast_value
        }
    }
}
```

which trivially satisfies the validity invariant but is crazy dangerous if you break the safety invariant.

Reply:
Yeah I'd prefer to use a type the user is most likely already familiar with.